seahorn/lib/array_list_helper.c

```diff
@@ -19,6 +19,28 @@ void initialize_array_list(struct aws_array_list *const list) {
   list->data = can_fail_malloc(list->current_size);
   list->alloc = sea_allocator();
 }
+
+void failed_allocation_after_initialize(struct aws_array_list *const list) {
+  if (list->data == NULL) {
+    list->current_size = 0;
+    list->length = 0;
+  }
+}
+
+void initialize_bounded_array_list(struct aws_array_list *const list,
+                                const size_t max_block_size) {
+  list->current_size = nd_size_t();
+  assume(list->current_size <= max_block_size);
+  list->item_size = nd_size_t();
+  list->length = nd_size_t();
+  assume(list->length <= list->current_size);
+  assume(list->length * list->item_size <= max_block_size);
+  assume(list->current_size >= list->length * list->item_size);
+  list->data = can_fail_malloc(list->current_size);
+  list->alloc = sea_allocator();
+  failed_allocation_after_initialize(list);
```

Comment:
better split early and assume that malloc does not fail. Early on, non-deterministically either set `data` to null, or call malloc.

The implementation of this functions for klee and seahorn will be different. For klee we want to be specific so that it works well. For seahorn, we want to be more abstract so that we test `is_valid` series of functions.